### PR TITLE
Add unit tests for 'restore.go' (TableRestore and chunkRestore)

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

Improve unit test coverage of restore package, part 2 of N.

### What is changed and how it works?

Add unit tests (29% → 40%).

Performed some function extraction to allow for more focused testing. 

### Check List 

Tests 

 - Unit test

Side effects

Related changes

### CODE REVIEW

**Comment by kennytm** [comment]
PTAL @lonng @july2993

File: lightning/restore/restore.go [hunk 1]
```diff
@@ -1607,42 +1600,65 @@ outside:
...
-		break
-	case <-ctx.Done():
-		return ctx.Err()
-	case deliverResult := <-deliverCompleteCh:
-		if deliverResult.err == nil {
-			panic("unexpected: deliverCompleteCh prematurely fulfilled with no error")
-		}
-		return errors.Trace(deliverResult.err)
+	err = send(deliveredKVs{kvs: nil, offset: lastOffset, rowID: lastRowID})
+	return
```
Comment (lonng) [inline lightning/restore/restore.go]: I think it's better to move L1621-L1622 to above L1590 is better and replace `break` with `return`

File: lightning/restore/restore.go [hunk 2]
```diff
@@ -1607,42 +1600,65 @@ outside:
...
+
+	kvsCh := make(chan deliveredKVs, maxKVQueueSize)
+	deliverCompleteCh := make(chan deliverResult)
+
+	go func() {
+		dur, err := cr.deliverLoop(ctx, kvsCh, t, engineID, dataEngine, indexEngine, rc)
+		deliverCompleteCh <- deliverResult{dur, err}
+	}()
+
+	defer close(kvsCh)
```
Comment (lonng) [inline lightning/restore/restore.go]: `deliverCompleteCh` should be closed also.

File: lightning/restore/restore_test.go [hunk 3]
```diff
@@ -76,3 +96,689 @@ func (s *restoreSuite) TestNewTableRestoreFailure(c *C) {
...
+	fakeDataDir := c.MkDir()
+	fakeDataFilesCount := 6
+	fakeDataFilesContent := []byte("INSERT INTO `table` VALUES (1, 2, 3);")
+	c.Assert(len(fakeDataFilesContent), Equals, 37)
+	fakeDataFiles := make([]string, 0, fakeDataFilesCount)
+	for i := 1; i <= fakeDataFilesCount; i++ {
+		fakeDataPath := path.Join(fakeDataDir, fmt.Sprintf("db.table.%d.sql", i))
+		err = ioutil.WriteFile(fakeDataPath, fakeDataFilesContent, 0644)
+		c.Assert(err, IsNil)
+		fakeDataFiles = append(fakeDataFiles, fakeDataPath)
```
Comment (lonng) [inline lightning/restore/restore_test.go]: All files should be deleted in `TearDownTest`.

File: lightning/restore/restore.go (hunk 1 above)
Comment (kennytm) [reply lightning/restore/restore.go]: No, line 1621-1623 will be hit by Line 1575 as well.

**Comment by GregoryIan** [comment]
LGTM

File: lightning/restore/restore_test.go (hunk 3 above)
Comment (kennytm) [reply lightning/restore/restore_test.go]: I'm going to move these into `SetUpSuite` since the data source is going to be immutable anyway.

Comment (kennytm) [reply lightning/restore/restore_test.go]: Also, `c.MkDir()` will return a [fresh directory](https://github.com/pingcap/check/blob/67f458068fc864dabf17e38d4d337f28430d13ed/check.go#L159) for every call, so the stale data from the previous tests won't affect the next ones anyway.

**Comment by kennytm** [comment]
@lonng PTAL again. Moved the close inside the goroutine, otherwise the `<-` may crash due to execution order.

**Pull request merged**